src/openfermion/resource_estimates/pbc/thc/utils/gvec_logic.py

```diff
@@ -0,0 +1,184 @@
+    """Define mapping momentum_transfer_map[Q][k1] = k2 that satisfies
+    k1 - k2 + G = Q.
+    where k1, k2, Q are all tuples of integers [[0, Nkx-1], [0, Nky-1],
+    [0, Nkz-1]]
+    and G is [{0, Nkx}, {0, Nky}, {0, Nkz}].
+
+    This is computed from `get_delta_kp_kq_Q` which computes k1 - k2 -Q = S.
+    Thus k1 - k2 = Q + S which shows that S is [{0, Nkx}, {0, Nky}, {0, Nkz}].
+
+    Thus to compute map[Q, k1] = k2
+
+    Args:
+      kmesh: kmesh [nkx, nky, nkz] the number of kpoints in each direction
+        scaled_kpts: miller index representation
+          [[0, nkx-1], [0, nky-1], [0, nkz-1]] of all the kpoints
+
+    Returns:
+      transfer map satisfying k1 - k2 + G = Q in matrix form map[Q, k1] = k2
+
+    """
+    nkpts = len(scaled_kpts)
+    delta_k1_k2_Q_int = get_delta_kp_kq_Q(scaled_kpts)
+    transfer_map = np.zeros((nkpts, nkpts), dtype=np.int32)
+    for kpidx, kqidx, qidx in itertools.product(range(nkpts), repeat=3):
+        # explicitly build my transfer matrix
+        if np.allclose(
+            [
+                np.rint(delta_k1_k2_Q_int[kpidx, kqidx, qidx][0]) % kmesh[0],
+                np.rint(delta_k1_k2_Q_int[kpidx, kqidx, qidx][1]) % kmesh[1],
+                np.rint(delta_k1_k2_Q_int[kpidx, kqidx, qidx][2]) % kmesh[2],
+            ],
+                0,
+        ):
+            transfer_map[qidx, kpidx] = kqidx
+    return transfer_map
+
+
+def build_conjugate_map(kmesh, scaled_kpts):
+    """build mapping such that map[k1] = -k1
+
+    Args:
+      kmesh: kpoint mesh
+      scaled_kpts: integer k-points
+
+    Returns:
+      kconj_map: conjugate k-point mapping
+
+    """
+    nkpts = len(scaled_kpts)
+    kpoint_dict = dict(
+        zip(
+            [tuple(map(int, scaled_kpts[x])) for x in range(nkpts)],
+            range(nkpts),
+        ))
+    kconj_map = np.zeros((nkpts), dtype=int)
+    for kidx in range(nkpts):
+        negative_k_scaled = -scaled_kpts[kidx]
+        fb_negative_k_scaled = tuple((
+            int(negative_k_scaled[0]) % kmesh[0],
+            int(negative_k_scaled[1]) % kmesh[1],
+            int(negative_k_scaled[2]) % kmesh[2],
+        ))
+        kconj_map[kidx] = kpoint_dict[fb_negative_k_scaled]
+    return kconj_map
+
+
+def build_G_vectors(kmesh):
+    """Build all 8 Gvectors
+
+    Args:
+      kmesh: returns tuple: G_dict a dictionary mapping miller index to
+        appropriate
+    G_vector index.  The actual cell Gvector can be recovered with
+    np.einsum("x,wx->w", (n1, n2, n3), cell.reciprocal_vectors()
+
+    Returns:
+      G_dict a dictionary mapping miller index to appropriate
+      G_vector index.  The actual cell Gvector can be recovered with
+      np.einsum("x,wx->w", (n1, n2, n3), cell.reciprocal_vectors()
+
+    """
+    G_dict = {}
+    indx = 0
+    for n1, n2, n3 in itertools.product([0, -1], repeat=3):
+        G_dict[(n1 * kmesh[0], n2 * kmesh[1], n3 * kmesh[2])] = indx
+        indx += 1
+    return G_dict
```

Comment:
this could probably be written with a dictionary comprehension and `enumerate`, although I don't think that will necessarily make it _more_ readable

Reply:
Yeah, I think I will leave it as is. This module was mostly for some sanity checks.